test/parallel/test-quic-send-fd.js

```diff
@@ -0,0 +1,70 @@
+'use strict';
+const common = require('../common');
+if (!common.hasQuic)
+  common.skip('missing quic');
+
+const assert = require('assert');
+const quic = require('quic');
+const fs = require('fs');
+
+const fixtures = require('../common/fixtures');
+const key = fixtures.readKey('agent1-key.pem', 'binary');
+const cert = fixtures.readKey('agent1-cert.pem', 'binary');
+const ca = fixtures.readKey('ca1-cert.pem', 'binary');
+
+const server = quic.createSocket({ port: 0, validateAddress: true });
+
+server.listen({
+  key,
+  cert,
+  ca,
+  rejectUnauthorized: false,
+  maxCryptoBuffer: 4096,
+  alpn: 'meow'
+});
+
+server.on('session', common.mustCall((session) => {
+  session.on('secure', common.mustCall((servername, alpn, cipher) => {
+    const stream = session.openStream({ halfOpen: false });
+    stream.sendFD(fs.openSync(__filename, 'r'));
+    stream.on('data', common.mustNotCall());
+    stream.on('finish', common.mustNotCall());
+    stream.on('close', common.mustCall());
+    stream.on('end', common.mustNotCall());
+  }));
+
+  session.on('close', common.mustCall());
+}));
+
+server.on('ready', common.mustCall(() => {
+  const client = quic.createSocket({
+    port: 0,
+    client: {
+      key,
+      cert,
+      ca,
+      alpn: 'meow'
+    }
+  });
+
+  const req = client.connect({
+    address: 'localhost',
+    port: server.address.port
+  });
+
+  req.on('stream', common.mustCall((stream) => {
+    const data = [];
+    stream.on('data', (chunk) => data.push(chunk));
+    stream.on('end', common.mustCall(() => {
+      assert.deepStrictEqual(Buffer.concat(data), fs.readFileSync(__filename));
+
+      // TODO(addaleax): Figure out why .close() is insufficient.
+      client.destroy();
+      server.destroy();
```

Comment:
Try moving the `client.close()` and `server.close()` to the `on('close')` event and see if that helps. On the end event, we can't be certain that all of the data has been transferred successfully (there still might be retransmissions, etc). The 'close' event tells us that the underlying `QuicStream` has fully completed. tl;dr destroying in end may prematurely cut off data transmission

Reply:
@jasnell So … that works, but I would have expected `.close()` to implicitly end all streams that haven’t been `.end()`ed yet.

> tl;dr destroying in end may prematurely cut off data transmission

There is no data transmission in this test that would be relevant here, though. The client doesn’t send anything and the server is done sending at this point.

Reply:
Yep, that's something we'll have to look at.

For `QuicSession`, `close()` initiates a graceful close, which is supposed to allow all currently open `QuicStream` instances to close and destroy before closing and destroying the `QuicSession`. Calling `close()` should prevent any new `QuicStream` instances from being created or accepted, but it does not tear down the `QuicSession` immediately so long as there is still an open `QuicStream`. So if having the `close()` here doesn't work, we need to revisit. 

Also, it's important that the internal `ngtcp2_conn` is not destroyed while inside one of the ngtcp2 callbacks.

Reply:
> So if having the `close()` here doesn't work, we need to revisit.

It does work, we just have to call `stream.end()` explicitly along with it.

(I’ll push that later together with other changes.)

Reply:
ah... ok that makes sense then. When using `sendFD()` the writable side of the JavaScript stream should be closed immediately without actually closing the stream. That should be straightforward to do.

Reply:
This is on the other side, not the one that uses `sendFD()`… Closing the writable side of the stream is already happening here

Would it be a bad idea to make `.close()` imply `.end()` if the stream has not been ended yet?

Reply:
that *should* be ok.